File: Baekjun/9.4/2616. 소형기관차/solution.swift
let N = Int(readLine()!)!
var A = readLine()!.split(separator: " ").compactMap { Int($0) }
let M = Int(readLine()!)!

for i in 1 ..< N {
    A[i] += A[i-1]
}

A.insert(0, at: 0)

var dp = Array(repeating: Array(repeating: 0, count: N + 1), count: 4)

for i in 1...3 {
    for j in i * M...N {
        dp[i][j] = max(dp[i][j-1], dp[i-1][j-M] + A[j]-A[j-M])
    }
}

print(dp.last!.last!)

File: Baekjun/9.4/2616. 소형기관차/solution.py
N = int(input())
A = list(map(int, input().split()))

for i in range(1, N):
    A[i] += A[i - 1]

A.insert(0, 0)
M = int(input())

# 길이가 M인 합이 가장 큰 연속된 수열 3개를 찾으면 된다.
dp = [[0] * (N + 1) for _ in range(4)]

for i in range(1, 4):
    for j in range(i * M, N + 1):
        dp[i][j] = max(dp[i][j - 1], dp[i - 1][j - M] + A[j] - A[j - M])

print(dp[-1][-1])
